app/questionnaire/navigator.py

```diff
@@ -1,193 +1,293 @@
+
+    def get_last_group_id(self):
+        return self.survey_json['groups'][-1]['id']
+
     def get_first_block_id(self):
         return self.survey_json['groups'][0]['blocks'][0]['id']
 
     def get_blocks(self):
         blocks = []
         for group in self.survey_json['groups']:
-            blocks.extend([block for block in group['blocks']])
+            blocks.extend([{
+                "group_id": group['id'],
+                "group_instance": 0,
+                "block": block,
+            } for block in group['blocks']])
+
+            if 'routing_rules' in group:
+                for rule in group['routing_rules']:
+                    if 'repeat' in rule.keys():
+                        no_of_times = evaluate_repeat(rule['repeat'], self.answer_store)
+                        for i in range(1, no_of_times):
+                            blocks.extend([{
+                                "group_id": group['id'],
+                                "group_instance": i,
+                                "block": block,
+                            } for block in group['blocks']])
         return blocks
 
-    @staticmethod
-    def get_first_location():
-        return Navigator.PRECEEDING_INTERSTITIAL_PATH[0]
+    @classmethod
+    def get_first_location(cls):
+        return cls.PRECEEDING_INTERSTITIAL_PATH[0]
 
-    def get_next_location(self, answers=None, current_location_id=None):
+    def get_next_location(self, current_group_id=None, current_block_id=None, current_iteration=0):
         """
         Returns the next 'location' to visit given a set of user answers
-        :param answers:
-        :param current_location_id:
+        :param current_group_id:
+        :param current_block_id:
+        :param current_iteration:
         :return:
         """
-        answers = answers or {}
-        location_path = self.get_location_path(answers)
-        if current_location_id in location_path:
-            current_location_index = location_path.index(current_location_id)
+        current_group_id = current_group_id or self.first_group_id
+        location_path = self.get_location_path()
+
+        this_block = {
+            "block_id": current_block_id,
+            "group_id": current_group_id,
+            "group_instance": current_iteration,
+        }
+
+        if this_block in location_path:
+
+            # Get blocks to be visited
+            block_ids = [index for (index, path_item) in enumerate(location_path) if path_item['block_id'] == current_block_id]
+
+            # Return the index of the block id to be visited
+            current_location_index = block_ids[current_iteration]
 
             if current_location_index < len(location_path) - 1:
                 return location_path[current_location_index + 1]
 
-    def get_previous_location(self, answers=None, current_location_id=None):
+    def get_previous_location(self, current_group_id=None, current_block_id=None, current_iteration=0):
```

Comment:
Lots of logic between get_previous_location and get_next_location is shared. We should refactor those bits